src/controllers/rendering/controllerrenderingengine.cpp

```diff
@@ -0,0 +1,394 @@
+    if (!m_quickWindow) {
+        m_waitCondition.wait(&m_mutex);
+    }
+    if (m_isValid) {
+        m_renderControl->prepareThread(m_pThread.get());
+    }
+}
+
+void ControllerRenderingEngine::requestSendingFrameData(
+        Controller* controller, const QByteArray& frame) {
+    emit sendFrameDataRequested(controller, frame);
+}
+
+void ControllerRenderingEngine::setup(std::shared_ptr<QQmlEngine> qmlEngine) {
+    DEBUG_ASSERT(QThread::currentThread() == thread());
+    QSurfaceFormat format;
+    format.setSamples(m_screenInfo.msaa);
+    format.setDepthBufferSize(16);
+    format.setStencilBufferSize(8);
+
+    m_context = std::make_unique<QOpenGLContext>();
+    m_context->setFormat(format);
+    VERIFY_OR_DEBUG_ASSERT(m_context->create()) {
+        kLogger.warning() << "Unable to initialize controller screen rendering. Giving up";
+        m_waitCondition.wakeAll();
+        return;
+    }
+    connect(m_context.get(),
+            &QOpenGLContext::aboutToBeDestroyed,
+            this,
+            &ControllerRenderingEngine::finish);
+
+    m_offscreenSurface = std::make_unique<QOffscreenSurface>();
+    m_offscreenSurface->setFormat(m_context->format());
+
+    VERIFY_OR_DEBUG_ASSERT(QMetaObject::invokeMethod(
+                                   qApp,
+                                   [this] {
+                                       m_offscreenSurface->create();
+                                   },
+                                   // This invocation will block the current thread!
+                                   Qt::BlockingQueuedConnection) &&
+            m_offscreenSurface->isValid()) {
+        kLogger.warning() << "Unable to create the OffscreenSurface for controller "
+                             "screen rendering. Giving up";
+        m_offscreenSurface.reset();
+        m_waitCondition.wakeAll();
+        return;
+    }
+
+    m_renderControl = std::make_unique<QQuickRenderControl>(this);
+    m_quickWindow = std::make_unique<QQuickWindow>(m_renderControl.get());
+
+    if (!qmlEngine->incubationController()) {
+        qmlEngine->setIncubationController(m_quickWindow->incubationController());
+    }
+
+    m_quickWindow->setGeometry(0, 0, m_screenInfo.size.width(), m_screenInfo.size.height());
+
+    m_isValid = true;
+    m_waitCondition.wakeAll();
+}
+
+void ControllerRenderingEngine::finish() {
+    disconnect(this);
+    m_isValid = false;
+
+    if (m_context && m_context->isValid()) {
+        disconnect(m_context.get(),
+                &QOpenGLContext::aboutToBeDestroyed,
+                this,
+                &ControllerRenderingEngine::finish);
+        m_context->makeCurrent(m_offscreenSurface.get());
+        m_renderControl.reset();
+
+        std::shared_ptr<QOffscreenSurface> pOffscreenSurface = std::move(m_offscreenSurface);
+        QMetaObject::invokeMethod(
+                qApp,
+                [pOffscreenSurface] {
+                    pOffscreenSurface->destroy();
+                });
```

Comment:
Ideally, yes. but making it blocking leads to a deadlock (remember the chain of destructor passed previously)

Reply:
I don't get it yet. So is `finish` always being called from the main thread then? Or is the main thread already locked?

Reply:
Not always, `finish` may be called by the `controller` thread, while the main thread is waiting for `controller` to shutdown. This deadlock prevent Mixxx from stopping, but doesn't not occur if you stop the controller from the preferences.

Reply:
Let me rephrase my question: In what thread is this method designed to run?

Reply:
The `ControllerRenderingEngine` thread. But the main thread [will be blocked](https://github.com/mixxxdj/mixxx/blob/a57332ad8e4eb9719b51ee7e1fdeb873035faca5/src/controllers/controllermanager.cpp#L134) till it completes when Mixxx is shutting down (as a reminder, this where I had the nasty `while` loop with Event Loop execution to prevent the deadlock in the version where `ControllerRenderingEngine` was ran in the GUI thread)

Reply:
I see. I really don't like this hidden assumption here that `controllerRendingEngine::finish` will only be called when the `ControllerManager` is shutting down which in tern assumes that it its running the main thread or rather the same thread that the fbo was created in. Do you see why I'm concerned over this brittleness?

Reply:
I don't know. I don't think so. The problem is not naming but architectural. As I said, lots of implicit assumptions that can easily break. Do you see why I'm concerned?

Reply:
Not really if I'm honest. `ControllerRenderingEngine` is private so calling will only happens locally. This means there shouldn't be any risk this gets randomly called from another thread in the future than `ControllerRenderingEngine`'s thread. 
I'm a bit confused on the architectural issues your are talking about here. OpenGL resources have thread affinity, and they need to be destructed within the thread they have been created into. Applying DRY here, I thought putting this procedure in the same place and calling it when needed was the best way forward, but I guess I might be missing something. Two options I can think of:
1. Put the destroy/teardown procedures in a macro so it gets called directly in the method
2. Add some verify assert in the `finish` method to ensure the thread affinity

Reply:
## Currently
_Mixxx exit sequence diagram_
![Sequence diagram](https://github.com/mixxxdj/mixxx/assets/7086688/7fa0c0a3-f26f-4f1c-a1ea-5daa519f229b)
_Controller stopped sequence diagram_
![Sequence diagram(1)](https://github.com/mixxxdj/mixxx/assets/7086688/08b29e3c-4bfb-4950-b1ea-01b20abc2ab9)
## Proposed

_Mixxx exit sequence diagram_
![Sequence diagram(2)](https://github.com/mixxxdj/mixxx/assets/7086688/729d9c4a-3d03-46b1-9f34-04322ecbcdc1)

_Controller stopped sequence diagram_
![Sequence diagram(3)](https://github.com/mixxxdj/mixxx/assets/7086688/6aa98523-15bc-4063-b4e5-3d58eae16a26)

Reply:
could we move the GL teardown/termination explicit and blocking in the stop rendering call instead of the destructor? (do we ever stop rendering if we're not tearing everything down).  I am imagining controllerManager->stop() that would block until all that was done, and then the destructors are *only* freeing up memory and not doing anything complex.

Reply:
It is actually already moved in `ControllerRenderingEngine::stop`, and called in `ControllerScriptEngineLegacy::shutdown`. The one which is done directly in the destructor is already in place, in `ControllerManager::~ControllerManager`. Happy to change that, but in another PR.
It's worth to mention that this will not solve the scheduling problem: if the main thread is waiting a thread, it isn't running the EventLoop so cannot handle a `BlockingQueue` call and will still deadlock. Only way I can think of is the spinning approach I had previously, but @Swiftb0y asked me to remove it.

Reply:
Thank you for the diagrams. they look very solid. The problem is that that same clarity is currently not visible as easily from the code itself. I think its totally fine for the dtors to be doing complex stuff (as long as its purely resource management, where resources could mean threads, connections, etc). 

> Add some verify assert in the finish method to ensure the thread affinity

Yes. that would at least make it possible to spot it in the logs when the assumptions over the threads are violated. 

>  if the main thread is waiting a thread, it isn't running the EventLoop so cannot handle a BlockingQueue call and will still deadlock.

Is it necessary for the main thread to wait for the Controller?

As I understand it right now, the controller threads currently do not block for surface destruction. Are there consequences to the order of destruction being undefined (surface before/after renderingthread GL).

Reply:
As a good rule of thumb, threads running an event loop should never be blocked if possible (which is kind of the implication of the "never spin the event loop manually" rule). 
I feel like there must be a solution for the problem of waiting QThread to join, without blocking the event loop in that thread. This doesn't feel like a unique problem.

Reply:
> Is it necessary for the main thread to wait for the Controller?

It is as otherwise, there is a risk of Mixxx process terminating before the controller have been properly shutdown (e.g running the `shutdown` method on mapping, which usually turns off all LED, and now also runs the splash off scene/clear the screen). 

> As I understand it right now, the controller threads currently do not block for surface destruction. Are there consequences to the order of destruction being undefined (surface before/after renderingthread GL)?

I am not 100% sure, but I will take the guess that there shouldn't be. My understanding of the (very incomplete) doc on the matter is the the GUI thread may have had allocated windows to be used as surfaces, depending of the platform allocation. They do not appear to be directly related the GL context so I'm going to take the risk and say it shouldn't be a problem as long as they don't get used after GL resources have been freed. As I have done all the dev and most of the testing on Linux tho, there could be bad surprises on other platforms.

Reply:
> This doesn't feel like a unique problem.

It definitely is not and not the first time I need top take my dirty spin  lock loop out. I which there would be some kind of "select", allowing you to keep handling even as they are coming in, while you are waiting for specific signal/thread to join

Reply:
that's an interesting idea -- maybe put it in the wiki and link it from the code?

Reply:
@Swiftb0y I have slightly refactored the teardown so `finish` is a pure slot and only get triggered by `QThread::finished`. This should help with readability.
Also, I did some kinda chaos testing by adding some `VERIFY_OR_TERMINATE` randomly triggered at various steps in the frame rendering cycle (different position in the code + different frame count). I managed to find one crash and added [the right signal handler for it](https://github.com/mixxxdj/mixxx/blob/ea12d82690991cdd19d42baed67acc26e188f190/src/controllers/scripting/legacy/controllerscriptenginelegacy.cpp#L554)
I have also done some more live testing on the S4 Mk3 and got a completely stable experience.

Reply:
yeah I can put it there

Reply:
does the image need any updating since the  notes?

Reply:
No, the sequence remain the same, I just changed the implementation. Thanks for that!